src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/UseAsSpanInsteadOfRangeIndexerTests.cs

```diff
@@ -0,0 +1,725 @@
+                    WithArguments("AsSpan", "System.Range", typeName + "[]"));
+        }
+
+        [Theory]
+        [MemberData(nameof(ArrayElementTypes))]
+        public static async Task ArrayToSpanReturn(string typeName)
+        {
+            await TestCS(@"
+using System;
+
+public class TestClass
+{
+    public Span<" + typeName + @"> TestMethod(" + typeName + @"[] input)
+    {
+        return {|CA1833:input[3..5]|};
+    }
+}");
+        }
+
+        [Theory]
+        [MemberData(nameof(ArrayElementTypes))]
+        public static async Task ArrayToSpanParameter(string typeName)
+        {
+            await TestCS(@"
+using System;
+
+public class TestClass
+{
+    private static int TestMethod2(Span<" + typeName + @"> input) => input.Length;
+
+    public int TestMethod(" + typeName + @"[] input)
+    {
+        return TestMethod2({|CA1833:input[3..5]|});
+    }
+}",
+                @"
+using System;
+
+public class TestClass
+{
+    private static int TestMethod2(Span<" + typeName + @"> input) => input.Length;
+
+    public int TestMethod(" + typeName + @"[] input)
+    {
+        return TestMethod2(input.AsSpan()[3..5]);
+    }
+}");
+        }
+
+        [Theory]
+        [MemberData(nameof(ArrayElementTypes))]
+        public static async Task ArrayToSpanCastLocal(string typeName)
+        {
+            await TestCS(@"
+using System;
+
+public class TestClass
+{
+    public int TestMethod(" + typeName + @"[] input)
+    {
+        Span<" + typeName + @"> tmp = (Span<" + typeName + @">)input[3..5];
+        return tmp.Length;
+    }
+}");
+        }
+
+        [Theory]
+        [MemberData(nameof(ArrayElementTypes))]
+        public static async Task ArrayToSpanCastReturn(string typeName)
+        {
+            await TestCS(@"
+using System;
+
+public class TestClass
+{
+    public Span<" + typeName + @"> TestMethod(" + typeName + @"[] input)
+    {
+        return (Span<" + typeName + @">)input[3..5];
+    }
+}");
+        }
+
+        [Theory]
+        [MemberData(nameof(ArrayElementTypes))]
+        public static async Task ArrayToSpanCastParameter(string typeName)
+        {
+            await TestCS(@"
+using System;
+
+public class TestClass
+{
+    private static int TestMethod2<T>(Span<T> input) => input.Length;
+
+    public int TestMethod(" + typeName + @"[] input)
+    {
+        return TestMethod2((Span<" + typeName + @">)input[3..5]);
+    }
+}");
+        }
+
+        [Theory]
+        [MemberData(nameof(ArrayElementTypes))]
+        public static async Task ArrayToReadOnlyMemoryLocal(string typeName)
+        {
+            await TestCS(@"
+using System;
+
+public class TestClass
+{
+    public int TestMethod(" + typeName + @"[] input)
+    {
+        ReadOnlyMemory<" + typeName + @"> tmp = input[3..5];
+        return tmp.Length;
+    }
+}",
+                VerifyCS.Diagnostic(UseAsSpanInsteadOfRangeIndexerAnalyzer.ArrayReadOnlyRule).
+                    WithLocation(8, 32 + typeName.Length).
```

Comment:
Can we convert all tests to be markup based?

Reply:
As far as I can tell, markup can't verify the string arguments.  So I have one of each kind of test verifying the arguments, and all of the other tests use markup.

Reply:
Ah got you. Thanks for the explanation. Maybe worth adding a comment so someone in future does not end up converting these to be markup based.